python/cudf_polars/cudf_polars/dsl/tracing.py

```diff
@@ -0,0 +1,53 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Tracing."""
+
+from __future__ import annotations
+
+import functools
+from typing import TYPE_CHECKING, Any
+
+import nvtx
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+
+    import cudf_polars.containers.dataframe
+
+
+CUDF_POLARS_NVTX_DOMAIN = "cudf_polars"
+
+
+def do_evaluate_with_tracing(
```

Comment:
I think it would be a bit more convenient to just pass in the IR object here, and then automatically wrap the `do_evaluate` classmethod (and automatically extract the name). E.g.:

```python
def do_evaluate_with_tracing(ir: IR) -> Callable[..., DataFrame]:
    cls = type(ir)

    @functools.wraps(cls.do_evaluate)
    def wrapped(*args: Any) -> DataFrame:
        with nvtx.annotate(
            message=cls.__name__,
            domain=CUDF_POLARS_NVTX_DOMAIN,
        ):
            return cls.do_evaluate(*args)

    return wrapped
```

Reply:
Hmm I'm a little confused how this could work. We have some methods like `RollingWindow.do_evaluate` that are regular methods, not classmethods.

I guess the difference is that those are `Expr` rather than `IR`? And we'll only deal with `IR` nodes here? If so, that simplifies things considerably (I had something that used instances of `ir` rather than `type(ir)` and that failed since `IR` nodes can't / shouldn't appear in the task graph.)

I'll give it a shot and will see if the tests pass.